File: day16/day16.py
import re
import sys
import os
sys.path.insert(1, os.path.join(sys.path[0], '..'))
import utils.utils as ut
from copy import deepcopy
import math


primes = ut.read_primes()


def valid(fields, ticket):
    total = 0
    for i in ticket:
        valid = False
        for f in fields:
            if int(i) >= int(f[1]) and int(i) <= int(f[2]): valid = True
            if int(i) >= int(f[3]) and int(i) <= int(f[4]): valid = True
        if valid == False:
            total += int(i)
            #print(i)
    return total


def part1(fields, ticket, tickets):
    total = 0
    for t in tickets:
        total += valid(fields, t)
    return total


def field_valid_for_ticket(f, t, i):
    #print('Test', f, t, i)
    if int(t[i]) >= int(f[1]) and int(t[i]) <= int(f[2]): return True
    if int(t[i]) >= int(f[3]) and int(t[i]) <= int(f[4]): return True
    return False


def field_valid_for_tickets(f, ts, i):
    count = 0
    for t in ts:
        if field_valid_for_ticket(f, t, i) == True: count += 1
    return count


def part2(fields, ticket, tickets):
    good = []
    for t in tickets:
        if valid(fields, t) == 0:
            good.append(t)

    # Valid tickets
    for g in good:
        print(g)   
    print(len(good))

    # Fields with ranges
    for f in fields:
        print(f)   
    print(len(fields))     

    # Which fields match all valid tickets in position i
    # ERROR: Most positions have several matches
    # ERROR: Some positions have none
    rules = {}
    for i in range(len(ticket)):
        count = []
        for f in fields:
            if field_valid_for_tickets(f, good, i) == len(good):
                #rules[i] = f
                count.append(f[0])
                #break
        print(i, count)

    for r in rules:            
        print(r, rules[r])
    print(len(rules))     

    print(ticket, len(ticket))
    result = 1
    for r in rules:
        #print(r, rules[r], rules[r][0])
        if rules[r][0].startswith('departure'):
            #print(r, ticket[r])
            result = result * int(ticket[r])


    return result


def solution(file):
    print("Input: ", file)
    lines = ut.read_lines(file)

    i = 0
    # Rules
    fields = []
    while len(lines[i]) > 0:
        m = re.match('^(.*):\s(\d+)-(\d+)\sor\s(\d+)-(\d+)$', lines[i])
        if m == None:
            break 
        fields.append([m.group(x) for x in range(1, 6)])
        #print([m.group(x) for x in range(1, 6)])
        i += 1

    i += 1
    # Ticket
    ticket = lines[i].split(',')
    #print(ticket)
    i += 2

    # Tickets
    tickets = []
    while i < len(lines) > 0:
        tickets.append(lines[i].split(','))
        #print(lines[i].split(','))
        i += 1

    #p1 = part1(fields, ticket, tickets)
    #print("Part1: ", p1)

    p2 = part2(fields, ticket, tickets)
    print("Part2: ", p2)


def main():
    #solution("day16-test.txt")
    #solution("day16-test2.txt")
    solution("day16-input.txt")


if __name__ == '__main__':
    main()

File: day16/day16.cpp
#include "utils.h"


struct Field
{
    string name;
    int    min1;
    int    max1;
    int    min2;
    int    max2;
};
using Fields  = vector<Field>;
using Ticket  = vector<int>;
using Tickets = vector<Ticket>;


bool matches_field(const Field& field, int value)
{
    if ((value >= field.min1) && (value <= field.max1)) return true;
    if ((value >= field.min2) && (value <= field.max2)) return true;
    return false;
}


bool matches_any_field(const Fields& fields, int value)
{
    for (const auto& field: fields)
        if (matches_field(field, value)) return true;
    return false;
}


int sum_invalid_fields(const Fields& fields, const Ticket& ticket)
{
    int result = 0;
    for (auto value: ticket)
    {
        if (!matches_any_field(fields, value))
        {
            result += value;
        }
    }
    return result;
}

int part1(const Fields& fields, const Tickets& nearby_tickets, const Ticket& ticket)
{
    int result = 0;
    for (const auto& nearby: nearby_tickets)
    {
        result += sum_invalid_fields(fields, nearby);
    }
    return result;
}


bool is_valid_ticket(const Fields& fields, const Ticket& ticket)
{
    for (auto value: ticket)
    {
        if (!matches_any_field(fields, value))
        {
            return false;
        }
    }
    return true;
}


bool matches_ticket(const Field& field, const Ticket& ticket, int index)
{
    return matches_field(field, ticket[index]);
}


bool matches_all_tickets(const Field& field, const Tickets& tickets, int index)
{
    for (const auto& ticket: tickets)
    {
        if (!matches_ticket(field, ticket, index))
            return false;
    }
    return true;
}


int part2(const Fields& fields, const Tickets& nearby_tickets, const Ticket& ticket)
{
    // Eliminate duff tickets.
    Tickets valid_tickets;
    for (const auto& nearby: nearby_tickets)
    {
        if (is_valid_ticket(fields, nearby))
            valid_tickets.push_back(nearby);
    }

    // Grid of fields and columns. All zeroes at first. A one means all tickets match 
    Tickets grid;
    for (const auto& field: fields)
    {
        Ticket row;
        for (int i = 0; i < fields.size(); ++i)
            row.push_back(matches_all_tickets(field, nearby_tickets, i));
    }

    // Now examine the grid to solve the constraints.

    return 0;
}


void solve(const string& filename)
{
    auto lines = aoc::read_lines(filename);
    lines.push_back(""); // Terminator
    size_t i = 0;

    Fields fields;
    while (true)
    {
        const auto& line = lines[i++];
        smatch sm;
        if (!regex_search(line, sm, regex{ R"((.*):\s+(\d+)-(\d+)\s+or\s+(\d+)-(\d+))" })) 
            break;
        fields.push_back(Field{sm[1], stoi(sm[2]), stoi(sm[3]), stoi(sm[4]), stoi(sm[5])});
    }

    Ticket ticket;
    auto items = aoc::split(lines[i++], ",", false);
    for (const auto& p: items)
        ticket.push_back(stoi(p));

    Tickets nearby_tickets;
    ++i;
    while (true)
    {
        const auto& line = lines[i++];
        if (line.size() == 0) break;
        auto items = aoc::split(line, ",", false);

        Ticket ticket;
        for (const auto& p: items)
            ticket.push_back(stoi(p));
        nearby_tickets.push_back(ticket);
    }

    cout << filename << '\n';
    cout << "Part1: " << part1(fields, nearby_tickets, ticket) << '\n';
    cout << "Part2: " << part2(fields, nearby_tickets, ticket) << '\n';
}


int main()
{
    solve("C:\\Projects\\aoc2020\\day16\\day16-input.txt");
}